src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -135,97 +143,88 @@ protected override void InvokeOnAttachmentStreamsReceived(int attachmentStreamCo
+                var minEtag = long.MaxValue;
+                foreach (ReplicationBatch replicationBatch in batches)
+                {
+                    cvs.Add(replicationBatch.LastAcceptedChangeVector);
+                    minEtag = Math.Min(minEtag, replicationBatch.LastEtagAccepted);
+                }
 
-            foreach (ShardedOutgoingReplicationHandler handler in _handlers)
-            {
-                if (handler.MissingAttachmentsInLastBatch)
-                    throw new MissingAttachmentException(handler.MissingAttachmentMessage);
+                _lastAcceptedChangeVector = ChangeVectorUtils.MergeVectorsDown(cvs);
+                _lastAcceptedEtag = minEtag;
             }
         }
 
-        private ReplicationBatch[] PrepareReplicationDataForShards(TransactionOperationContext context, IncomingReplicationHandler.DataForReplicationCommand dataForReplicationCommand)
+        private ReplicationBatches PrepareReplicationDataForShards(TransactionOperationContext context, DataForReplicationCommand dataForReplicationCommand)
         {
             DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Optimization possibility: instead of iterating over materialized batch, we can do it while reading from the stream");
-
-            var batches = new ReplicationBatch[_parent.Context.ShardCount];
-            for (int i = 0; i < _parent.Context.ShardCount; i++)
-            {
-                batches[i] = new();
-            }
-
+            var replicationBatches = new ReplicationBatches(this);
+            var batches = replicationBatches.Batches;
+            
             foreach (var item in dataForReplicationCommand.ReplicatedItems)
             {
                 int shard = GetShardNumberForReplicationItem(context, item);
                 batches[shard].Items.Add(item);
-                if (item is AttachmentReplicationItem attachmentReplicationItem)
+                if (item is AttachmentReplicationItem attachmentReplicationItem && attachmentReplicationItem.Stream == null)
                 {
-                    batches[shard].Attachments ??= new(SliceComparer.Instance);
-                    batches[shard].Attachments.Add(attachmentReplicationItem.Base64Hash, null);
+                    batches[shard].AttachmentStreams ??= new(SliceComparer.Instance);
+                    batches[shard].AttachmentStreams.TryAdd(attachmentReplicationItem.Base64Hash, null);
                 }
             }
 
             if (dataForReplicationCommand.ReplicatedAttachmentStreams != null)
             {
                 foreach (var attachment in dataForReplicationCommand.ReplicatedAttachmentStreams)
                 {
+                    if (attachment.Value.Stream is StreamsTempFile.InnerStream innerStream == false)
+                        throw new NotSupportedException("Cannot understand how to read stream: " + attachment.Value.Stream);
```

Comment:
I'm assuming that this cannot happen?
Message should be then:
```
$"Cannot understand how to read stream of type '{attachment.Value.Stream.GetType().Name}'. Should not happen!" 
```
?

Reply:
Done